### PR TITLE
Custom formatting via config.toml

### PR DESCRIPTION
**Pull request by dotaxis**
This allows customizing the output of spotifatius using the config.toml.

Includes `{status}` var which will print a play/pause unicode character depending on if the track is playing or paused.

Open to any suggestions on how to make it better, just let me know.

Also address https://github.com/AndreasBackx/spotifatius/issues/5

### CODE REVIEW

File: src/commands/monitor.rs
Comment (AndreasBackx) [inline src/commands/monitor.rs]: Could you move this to `OutputFormatter`? If the replacing is done on all `Output.text`, we could make the other messages below configurable as well.

Additionally, the tooltip can be configurable.

You'll need to add a `template_fields: HashMap<String, String>` or something similar where you pass what needs replacing. Otherwise the `OutputFormatter` won't know.

File: src/shared/config.rs
Comment (AndreasBackx) [inline src/shared/config.rs:15]: What do you think about having some extra formatting options?

```rust
pub struct TemplateConfigs {
	#[serde(default = "default_text_template")]
	pub text: TemplateConfig,
	#[serde(default = "default_tooltip_template")]
	pub tooltip: TemplateConfig,
}

pub struct TemplateConfig {
	pub default: String,
	pub playing: String,
	pub stopped: String,
	pub paused: String,
	pub added: String,
	pub removed: String,
}
```

This would allow configuration of icons and more so we do not hardcode those 2 unicode characters.

**Review by AndreasBackx — CHANGES_REQUESTED** [review]
This looks great, thank you! I think I primarily want to move the replace logic to the `OutputFormatter`. It would be nice if we additionally didn't hardcode the unicode icons. I've left a possible solution for that as well that would be relatively simple, even if a bit boilerplate-y.

**Comment by dotaxis** [comment]
I'll take a look at this today and see what I can come up with.

File: src/commands/output.rs
Comment (dotaxis) [inline src/commands/output.rs:70]: I can't get these blocks to trigger. It seems the `status` never seems to be `TrackStatus::Added` or `TrackStatus::Removed` even if I put them at the start of the `if` instead of down here. Spotify always opens to the last playing track and sends us that information instead.

Can we remove this block and the next one or is it just my machine that doesn't work?

**Comment by dotaxis** [comment]
I've done a bit more work on this. If you like the direction it's going I can add tooltip customization, but would rather wait to see if you want it refactored at all first.

I also started a review regarding the TrackStatus::Added/Removed triggers because I don't think they work.

**Comment by AndreasBackx** [comment]
Thank you! I'll try to give this a review this evening after work.

File: src/commands/output.rs
Comment (AndreasBackx) [inline src/commands/output.rs]:
```suggestion
            let is_liked = response.is_liked.unwrap_or_default();
            if is_liked {
```

File: src/commands/output.rs
Comment (AndreasBackx) [inline src/commands/output.rs]:
```suggestion
                    let liked_icon = if is_liked {
                    	&text_template.liked
                    } else {
                    	"" // Display no icon if not liked
                    }
```

The downside here is that you won't be able to control padding if it's liked or not. Though this is better than nothing for now. While I prefer  what I mentioned in my other comments, I think you're right in that it can be worked towards.

File: src/commands/output.rs
Comment (AndreasBackx) [inline src/commands/output.rs:30]: This is somewhat blurring the lines of the purposes of `OutputFormatter`, as it should exclusively format the `Output` struct. Though I'd need to give it some thought on how to make it better. I don't want to block your changes for that.

File: README.md
Comment (AndreasBackx) [inline README.md:100]:
```suggestion
🗒 **Note:** These formatting options are not yet final and might change in upcoming versions. They won't be considered breaking changes as a result, though a minor version bump will be done if this changes. [More info.](https://github.com/AndreasBackx/spotifatius/pull/8)

The JSON output can be formatted using the `format` var in config.toml:
```

Let's just say that this might change a bit still. I'm happy with your contributions, though I don't want to have to deal with people being annoyed in the future when it might change. Though it'll be done with care regardless.

**Review by AndreasBackx — CHANGES_REQUESTED** [review]

**Comment by AndreasBackx** [comment]
Forgot to leave a final comment with the review.

While I think the config can be improved, this is an improvement and I'm happy to extend it so it's better for you to use. I left some suggestions and with those applied, I think we can merge this.

Also please add an "Added" section to the Unreleased section of `CHANGELOG.md`. I'll put up a new version once it's merged and handle it appropriately. See https://keepachangelog.com/en/1.1.0/ for more info.

File: src/commands/output.rs
Comment (dotaxis) [reply src/commands/output.rs:30]: I'm pretty new to Rust so I would be very interested to see how you can improve this. I didn't exactly think copy-pasting the code from `monitor.rs` was a great solution, but it did work and was an improvement.

I guess we'll consider this resolved for now.

File: src/commands/output.rs
Comment (dotaxis) [reply src/commands/output.rs]: The padding can be controlled via a combination of `format` and `text_template` settings, although it's a bit unintuitive. Here's an example:

```toml
format = "{status}{liked}{title} {separator} {artist}"
[text_template]
playing = "|> "
paused = "|| "
liked = "+ "
```

File: src/commands/output.rs
Comment (dotaxis) [reply src/commands/output.rs:70]: @AndreasBackx can you take a look at this?

**Comment by AndreasBackx** [comment]
> Also please add an "Added" section to the Unreleased section of CHANGELOG.md. I'll put up a new version once it's merged and handle it appropriately. See [keepachangelog.com/en/1.1.0](https://keepachangelog.com/en/1.1.0/) for more info.

Should be good to merge after this.

**Comment by dotaxis** [comment]
Still waiting on some feedback on the Added/Removed logic not triggering

**Comment by AndreasBackx** [comment]
@dotaxis I'm unsure if you're referring to feedback from me. You mentioned the below earlier but I'm not sure what you mean with that review.

> I also started a review regarding the TrackStatus::Added/Removed triggers because I don't think they work.

Let me know if you need any of my help.

**Comment by dotaxis** [comment]
> @dotaxis I'm unsure if you're referring to feedback from me. You mentioned the below earlier but I'm not sure what you mean with that review.
> 
> > I also started a review regarding the TrackStatus::Added/Removed triggers because I don't think they work.
> 
> Let me know if you need any of my help.

I forgot to submit my review so only I could see it. Oops.

File: src/commands/output.rs
Comment (AndreasBackx) [reply src/commands/output.rs:70]: You trigger this by using `spotifatius toggle-liked` to add/remove the currently playing song to your Spotify liked library.

Comment (dotaxis) [reply src/commands/output.rs:70]: Oh, my bad. I thought this was supposed to be triggered by liking/removing a song via the spotify GUI.

Comment (AndreasBackx) [reply src/commands/output.rs:70]: Unfortunately Spotify does not push an update when that happens and it doesn't expose this on the dbus. One could technically implement the behaviour via continual polling on their API, though that is likely not gonna fly with rate limiting.

**Comment by dotaxis** [comment]
> > Also please add an "Added" section to the Unreleased section of CHANGELOG.md. I'll put up a new version once it's merged and handle it appropriately. See [keepachangelog.com/en/1.1.0](https://keepachangelog.com/en/1.1.0/) for more info.
> 
> Should be good to merge after this.

Sorry for the delay, been busy all week. Will get this done over the weekend.

**Comment by AndreasBackx** [comment]
No rush at all, take all the time you need. You're the primary user for it.

**Comment by dotaxis** [comment]
Okay, I added a line in the changelog. Let me know if you need anything else done.

File: CHANGELOG.md
Comment (AndreasBackx) [inline CHANGELOG.md]:
```suggestion
- Added custom formatting options via config.toml. #8 by @dotaxis 
```

File: src/commands/output.rs
Comment (AndreasBackx) [inline src/commands/output.rs:32]: You're not closing this properly: 
```
error: unexpected closing delimiter: `}`
   --> src/commands/output.rs:120:1
    |
42  |                 (Some(artist), Some(title)) => {
    |                                                - this delimiter might not be properly closed...
...
53  |                     };
    |                     - ...as it matches this but it has different indentation
...
120 | }
    | ^ unexpected closing delimiter

error: could not compile `spotifatius` (bin "spotifatius") due to 1 previous error
```

**Review by AndreasBackx — CHANGES_REQUESTED** [review]
Thanks, though make sure to test that they work. 😅 It does not compile right now.

**Comment by dotaxis** [comment]
There was an extra closing brace in the change you requested on line 52.

**Comment by AndreasBackx** [comment]
Thanks again! It might take me some time to create a new version however. You should be able to use it yourself in the meantime however.

**Comment by dom1torii** [comment]
Not sure where to ask, here or create a new issue but status icons don't display on my end. Do i need some font to display them or add something to config? Trying to use it with waybar and default module config

**Pull request merged**
